src/algorithms/heuristics/heuristics.cpp

```diff
@@ -132,6 +132,10 @@ template <class T> T basic(const Input& input, INIT init, float lambda) {
           continue;
         }
 
+        if (!current_r.is_valid_addition_for_max_tasks(input, job_rank)) {
```

Comment:
I'd be in favor of performing an inline explicit check here based on vehicle `max_number_of_tasks` and size of `current_r` and remove the `RawRoute::is_valid_addition_for_max_tasks` member altogether.

Having the test logic factored in a single function sounds appealing in the first place but:
- we still need to use explicit conditions elsewhere (cf the filters already applied for local search operators);
- an explicit test like `v.max_number_of_tasks == current_r.size()` (for single jobs) feels as expressive as a function call.

Reply:
In most cases where `RawRoute::is_valid_addition_for_max_tasks` method is used, type of job is not filted (can be both `JOB_TYPE::SINGLE` and `JOB_TYPE::PICKUP`)

Reply:
Yes, I noticed that. The test could then rely on the `is_pickup` value which is defined just below. In some other places for this file there is an `if` branching on the job type so the appropriate test for each case could be added there.

Reply:
Was busy with work, removed `RawRoute::is_valid_addition_for_max_tasks` and added `max_number_of_tasks` to docs/API.md